Hw03 frequency analysis

BBeRsErKeRR opened the pull request:
## Домашнее задание №3 «Частотный анализ»

### Чек-лист студента ([Что это?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0))
- [x] Я перечитал задание после решения. [Зачем перечитывать задание?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%82%D1%8B%D0%B2%D0%B0%D1%82%D1%8C-%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5)
- [x] Я запустил `go mod tidy`.
- [x] Я удалил `.sync` файл. [Зачем его удалять?](https://github.com/OtusGolang/home_work/wiki/%5B%D0%A1%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0%D0%BC%5D-%D0%9F%D1%80%D0%BE%D1%86%D0%B5%D1%81%D1%81-%D1%81%D0%B4%D0%B0%D1%87%D0%B8-%D0%94%D0%97#user-content-%D0%92%D0%B0%D1%80%D0%B8%D0%B0%D0%BD%D1%82-2)

### Критерии оценки
- [ ] Пайплайн зелёный - 4 балла
- [ ] Добавлены новые юнит-тесты - до 4 баллов
- [ ] Понятность и чистота кода - до 2 баллов

#### Зачёт от 7 баллов

BBeRsErKeRR committed:
feat(hw03): add top10 generato

BBeRsErKeRR committed:
fix(hw03): update depth

dshemin commented on hw03_frequency_analysis/top.go:
```diff
@@ -1,6 +1,41 @@
+)
+
+func Top10(input string) []string {
+	if input != "" {
```
Важно:

Более чистым и правильным решением будет проверить равна ли переменная пустой строке и вернуть nil из функции и потом обрабатывать строку, чем так. Появляется излишняя вложенность.

dshemin commented on hw03_frequency_analysis/top.go:
```diff
@@ -1,6 +1,41 @@
+	if input != "" {
+		// Generate base structs
+		words := make(map[string]int)
+		re := regexp.MustCompile(`\s+(-\s+)*`)
```
Совет:

Компиляция регулярок - дорогой процесс, Если регулярки статичные, то лучше их компиляцию выносить в приватные переменные пакета.

dshemin commented on hw03_frequency_analysis/top.go:
```diff
@@ -1,6 +1,41 @@
+			}
+		}
+		// Check if words more then 10
+		if len(words) >= 10 {
```
Критично:

Наша функция не будет работать если у нас в предложении меньше 10 слов? Если обратите внимание на пример в задании, там выводится топ 7 слов.

BBeRsErKeRR replied:
Внес исправления

dshemin commented on hw03_frequency_analysis/top.go:
```diff
@@ -1,6 +1,41 @@
+		// Generate base structs
+		words := make(map[string]int)
+		re := regexp.MustCompile(`\s+(-\s+)*`)
+		replace := regexp.MustCompile(`[,.!:"']+`)
```
Важно:

Для простых реплейсов, а у нас тут простая замена, лучше использовать методы из пакета strings. Более эффективное решение.

BBeRsErKeRR replied:
Предполагалось дополнительно решить задачку с апострофами, но пакет `"regexp"` не поддерживает lookahead/lookbehind.
Переписал на replacer.

dshemin commented on hw03_frequency_analysis/top_test.go:
```diff
@@ -7,7 +7,7 @@ import (
 
 // Change to true if needed.
-var taskWithAsteriskIsCompleted = false
+var taskWithAsteriskIsCompleted = true
```
Критично:

Нет дополнительных юнит-тестов, это -4 бала. Добавьте тесты на краевые случаи, например если у нас предложение всего из одного слова.

BBeRsErKeRR replied:
Добавил доп unit тесты

BBeRsErKeRR committed:
fix(hw03): resolve discuss

- add unit tests;
- migrate replacres to package level;
- fast return nil if iput empty string.

BBeRsErKeRR committed:
fix(hw03): avoid lint problems
